### PR TITLE
feat: type deserialization + encoding

### PR DESCRIPTION
**Pull request by iammadab**

### CODE REVIEW

File: src/contract/types.rs [hunk 1]
```diff
@@ -35,70 +42,140 @@ pub fn encode_document_field_type(
...
+            let value_as_i64: i64 = value_as_integer
                 .try_into()
                 .map_err(|_| Error::CorruptedData(String::from("expected integer value")))?;
-            Ok(Some(value_as_u64.to_be_bytes().to_vec()))
+
+            encode_integer(value_as_i64)
         }
         DocumentFieldType::Float => {
             let value_as_float = value.as_float().ok_or(field_type_match_error)?;
             Ok(Some(value_as_float.to_be_bytes().to_vec()))
```
Comment (shumkov) [inline src/contract/types.rs:53]: Don't we need to encode floats too?

File: src/contract/types.rs [hunk 2]
```diff
@@ -35,70 +42,140 @@ pub fn encode_document_field_type(
...
         }
+        DocumentFieldType::ByteArray => {
+            // Byte array could either be raw bytes or encoded as a base64 string
+            if value.is_text() {
```
Comment (shumkov) [inline src/contract/types.rs:57]: ByteArray actually always will be raw bytes

File: src/contract/types.rs [hunk 3]
```diff
@@ -35,70 +42,140 @@ pub fn encode_document_field_type(
...
+            }
+        }
+        DocumentFieldType::Date => {
+            let date_string = value.as_text().ok_or(field_type_match_error)?;
+            let date_as_integer: i64 = date_string
+                .parse()
+                .map_err(|_| Error::CorruptedData(String::from("invalid integer string")))?;
+            encode_integer(date_as_integer)
+        }
+        DocumentFieldType::Object => Ok(Some(Vec::new())),
```
Comment (shumkov) [inline src/contract/types.rs:84]: We don't index objects so we probably should just throw an error

File: src/contract/types.rs (hunk 2 above)
Comment (iammadab) [reply src/contract/types.rs:57]: Currently we test by converting json to cbor and byte arrays are represented as base64 in the json files, it's temporary tho will eventually move the base64 decoding to conversion function.

File: src/contract/types.rs (hunk 3 above)
Comment (iammadab) [reply src/contract/types.rs:84]: True, thanks.

File: src/contract/types.rs (hunk 2 above)
Comment (QuantumExplorer) [reply src/contract/types.rs:55]: yes @shumkov, this is basically just to allow tests to be much easier.

File: src/contract/types.rs [hunk 4]
```diff
@@ -35,70 +40,139 @@ pub fn encode_document_field_type(
...
+
+    // Flip the sign bit
+    // to deal with interaction between the domains
+    // 2's complement values have the sign bit set to 1
+    // this makes them greater than the positive domain in terms of sort order
+    // to fix this, we just flip the sign bit
+    // so positive integers have the high bit and negative integers have the low bit
+    // the relative order of elements in each domain is still maintained, as the
+    // change was uniform across all elements
+    wtr[0] = wtr[0] ^ 0b1000_0000;
```
Comment (QuantumExplorer) [inline src/contract/types.rs:105]: or just wtr[0] ^= 0b1000_0000;

File: src/contract/types.rs (hunk 2 above)
Comment (shumkov) [reply src/contract/types.rs:58]: Ok, sounds good. Should we add a TODO there to do not forget?

Comment (QuantumExplorer) [reply src/contract/types.rs:58]: I'm not sure we need to, what's the point of not keeping this?

**Review by QuantumExplorer — APPROVED** [review]

**Pull request merged**